File: src/harness/anthropic_llm.py
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any, Dict, List

from .llm import BaseLLM, build_system_prompt

try:
    import anthropic
except ImportError:
    anthropic = None  # type: ignore[assignment]


class AnthropicLLM(BaseLLM):
    """LLM backend that delegates to the Anthropic Messages API."""

    def __init__(
        self,
        *,
        api_key: str | None = None,
        model: str = "claude-sonnet-4-20250514",
        tool_schemas: list[dict[str, Any]] | None = None,
    ) -> None:
        super().__init__()
        if anthropic is None:
            raise ImportError(
                "The 'anthropic' package is required. Install it with: pip install anthropic"
            )
        self.model = model
        self.tool_schemas = tool_schemas or []
        self.api_key = self._resolve_api_key(api_key)
        self._client = anthropic.Anthropic(api_key=self.api_key)

    def set_tool_schemas(self, schemas: list[dict[str, Any]]) -> None:
        """Update tool schemas after initialization."""
        self.tool_schemas = schemas

    def generate(self, working_memory: Dict[str, Any]) -> Dict[str, Any]:
        tool_names = [s["name"] for s in self.tool_schemas]
        system_prompt = build_system_prompt(tool_names)
        messages = self._build_messages(working_memory)

        kwargs: dict[str, Any] = {
            "model": self.model,
            "system": system_prompt,
            "messages": messages,
            "max_tokens": 4096,
        }
        if self.tool_schemas:
            kwargs["tools"] = self.tool_schemas

        if self.on_token:
            return self._generate_streaming(kwargs)

        response = self._client.messages.create(**kwargs)
        return self._parse_response(response)

    def _generate_streaming(
        self,
        kwargs: dict[str, Any],
    ) -> Dict[str, Any]:
        """Stream tokens via on_token callback, return parsed final response."""
        with self._client.messages.stream(**kwargs) as stream:
            for event in stream:
                if event.type == "content_block_delta":
                    delta = event.delta
                    if hasattr(delta, "text") and self.on_token:
                        self.on_token(delta.text)

            final_message = stream.get_final_message()
        return self._parse_response(final_message)

    @staticmethod
    def _build_messages(
        working_memory: Dict[str, Any],
    ) -> List[Dict[str, Any]]:
        history: List[Dict[str, Any]] = working_memory.get("history", [])
        schema_feedback = working_memory.get("schema_feedback")

        # Build the initial user message with goal + context + summary
        first_parts = []
        if working_memory.get("goal"):
            first_parts.append(f"Goal: {working_memory['goal']}")
        if working_memory.get("context"):
            first_parts.append(
                f"Context: {json.dumps(working_memory['context'], ensure_ascii=False)}"
            )
        if working_memory.get("summary_memory"):
            first_parts.append(f"Summary: {working_memory['summary_memory']}")

        # Bug B fix: merge schema_feedback into initial message when history is empty
        if not history and schema_feedback:
            first_parts.append(
                f"Schema feedback: {json.dumps(schema_feedback, ensure_ascii=False)}"
            )

        first_msg = "\n".join(first_parts) if first_parts else json.dumps(
            working_memory, ensure_ascii=False, indent=2
        )

        messages: List[Dict[str, Any]] = [{"role": "user", "content": first_msg}]

        for turn in history:
            action = turn.get("action", {})
            action_type = action.get("action_type")

            if action_type == "tool_call":
                # Assistant decided to call a tool
                assistant_content = json.dumps(
                    {
                        "tool_name": action.get("tool_name"),
                        "arguments": action.get("arguments"),
                    },
                    ensure_ascii=False,
                )
                # Bug A fix: insert bridging user message if last message is also assistant
                if messages and messages[-1]["role"] == "assistant":
                    messages.append({"role": "user", "content": "Acknowledged. Continue."})
                messages.append({"role": "assistant", "content": assistant_content})

                # Bug C fix: use structured tool_result data with json.dumps
                tool_result_data = turn.get("tool_result")
                if tool_result_data and isinstance(tool_result_data, dict):
                    content = json.dumps(tool_result_data, ensure_ascii=False, default=str)
                else:
                    content = str(turn.get("observation", ""))
                messages.append({"role": "user", "content": content})

            elif action_type == "final_response":
                # Bug A fix: insert bridging user message if last message is also assistant
                if messages and messages[-1]["role"] == "assistant":
                    messages.append({"role": "user", "content": "Acknowledged. Continue."})
                messages.append(
                    {"role": "assistant", "content": action.get("content") or ""}
                )

        # If history was non-empty and last message is assistant, add continuation prompt
        if history and messages[-1]["role"] == "assistant":
            continuation = "Continue with the next step."
            if schema_feedback:
                continuation = (
                    f"Schema feedback: "
                    f"{json.dumps(schema_feedback, ensure_ascii=False)}\n"
                    f"Please try again."
                )
            messages.append({"role": "user", "content": continuation})

        return messages

    @staticmethod
    def _parse_response(response: Any) -> Dict[str, Any]:
        result: Dict[str, Any] = {}
        text_content: str | None = None
        for block in response.content:
            if block.type == "tool_use":
                result = {
                    "type": "tool_call",
                    "tool_name": block.name,
                    "arguments": block.input,
                }
                break  # tool_use takes priority
            if block.type == "text":
                text_content = block.text

        if not result and text_content is not None:
            result = {"type": "final_response", "content": text_content}
        elif not result:
            raise ValueError(
                "Anthropic response contained no usable content blocks."
            )

        # Extract usage information
        if hasattr(response, "usage") and response.usage:
            result["_usage"] = {
                "input_tokens": response.usage.input_tokens,
                "output_tokens": response.usage.output_tokens,
                "total_tokens": (
                    response.usage.input_tokens
                    + response.usage.output_tokens
                ),
            }
        return result

    @staticmethod
    def _resolve_api_key(explicit_key: str | None) -> str:
        key = explicit_key or os.environ.get("ANTHROPIC_API_KEY", "").strip()
        if key:
            return key

        env_path = Path(__file__).resolve().parents[2] / ".env"
        if env_path.exists():
            for line in env_path.read_text(encoding="utf-8").splitlines():
                stripped = line.strip()
                if not stripped or stripped.startswith("#") or "=" not in stripped:
                    continue
                name, value = stripped.split("=", 1)
                if name.strip() == "ANTHROPIC_API_KEY":
                    return value.strip().strip('"').strip("'")

        raise ValueError(
            "Anthropic API key not found. Set ANTHROPIC_API_KEY "
            "environment variable or pass api_key explicitly."
        )

File: ts/src/anthropic-llm.ts
/**
 * AnthropicLLM — Anthropic Messages API adapter with native tool_use support.
 */

import type { ToolSchema } from "./types.js";
import { BaseLLM, buildSystemPrompt } from "./llm.js";
import * as fs from "node:fs";
import * as path from "node:path";

// Re-export buildSystemPrompt for backward compatibility with tests
export { buildSystemPrompt };

// --- Retry helper ---

export function isRetryable(err: Error): boolean {
  const name = err.name;
  if (name === "APIConnectionError" || name === "APITimeoutError") return true;
  if (name === "RateLimitError") return true;
  if (name === "APIStatusError" && "status_code" in err) {
    return (err as any).status_code >= 500;
  }
  return false;
}

// --- Message type ---

interface Message {
  role: "user" | "assistant";
  content: string | any[];
}

// --- AnthropicLLM ---

export class AnthropicLLM extends BaseLLM {
  apiKey: string;
  model: string;
  toolSchemas: ToolSchema[];
  extraInstructions: string;
  declare onToken?: (token: string) => void;
  private _client: any;

  constructor(options?: {
    apiKey?: string;
    model?: string;
    baseUrl?: string;
    toolSchemas?: ToolSchema[];
  }) {
    super();
    this.model = options?.model ?? "claude-sonnet-4-20250514";
    this.toolSchemas = options?.toolSchemas ?? [];
    this.extraInstructions = "";
    this.apiKey = AnthropicLLM._resolveApiKey(options?.apiKey);

    // Dynamically import Anthropic SDK
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const Anthropic = require("@anthropic-ai/sdk").default;
      const clientOpts: Record<string, unknown> = { apiKey: this.apiKey };
      if (options?.baseUrl) {
        clientOpts.baseURL = options.baseUrl;
      }
      this._client = new Anthropic(clientOpts);
    } catch {
      throw new Error(
        "The '@anthropic-ai/sdk' package is required. Install it with: npm install @anthropic-ai/sdk",
      );
    }
  }

  setToolSchemas(schemas: ToolSchema[]): void {
    this.toolSchemas = schemas;
  }

  async generate(workingMemory: Record<string, unknown>): Promise<Record<string, unknown>> {
    const toolNames = this.toolSchemas.map((s) => s.name);
    const systemPrompt = buildSystemPrompt(toolNames, {
      nativeToolUse: true,
      extraSystemInstructions: this.extraInstructions,
    });
    const messages = AnthropicLLM._buildMessages(workingMemory);

    const kwargs: Record<string, unknown> = {
      model: this.model,
      system: systemPrompt,
      messages,
      max_tokens: 4096,
    };
    if (this.toolSchemas.length > 0) {
      kwargs.tools = this.toolSchemas;
    }

    if (this.onToken) {
      return this._callWithRetry(() => this._generateStreaming(kwargs));
    }

    return this._callWithRetry(async () => {
      const response = await this._client.messages.create(kwargs);
      return AnthropicLLM._parseResponse(response);
    });
  }

  private async _callWithRetry(fn: () => Promise<Record<string, unknown>>): Promise<Record<string, unknown>> {
    const maxRetries = 2;
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        return await fn();
      } catch (err: any) {
        if (isRetryable(err) && attempt < maxRetries) {
          const wait = 2 ** attempt;
          process.stderr.write(
            `\x1b[2m⟳ API 请求失败，${wait}s 后重试... (${err})\x1b[0m\n`,
          );
          await new Promise((resolve) => setTimeout(resolve, wait * 1000));
          continue;
        }
        throw new Error(`LLM API call failed: ${err.message ?? err}`, { cause: err });
      }
    }
    // unreachable
    throw new Error("Unexpected retry loop exit");
  }

  private async _generateStreaming(kwargs: Record<string, unknown>): Promise<Record<string, unknown>> {
    const stream = this._client.messages.stream(kwargs);
    for await (const event of stream) {
      if (
        event.type === "content_block_delta" &&
        event.delta?.text &&
        this.onToken
      ) {
        this.onToken(event.delta.text);
      }
    }
    const finalMessage = await stream.finalMessage();
    return AnthropicLLM._parseResponse(finalMessage);
  }

  static _buildMessages(workingMemory: Record<string, unknown>): Message[] {
    const history = (workingMemory.history ?? []) as Array<Record<string, any>>;
    const schemaFeedback = workingMemory.schema_feedback;

    // Build the initial user message with goal + context + summary
    const firstParts: string[] = [];
    if (workingMemory.goal) {
      firstParts.push(`Goal: ${workingMemory.goal}`);
    }
    if (workingMemory.context) {
      firstParts.push(`Context: ${JSON.stringify(workingMemory.context)}`);
    }
    if (workingMemory.summary_memory) {
      firstParts.push(`Summary: ${workingMemory.summary_memory}`);
    }

    // Bug B fix: merge schema_feedback into initial message when history is empty
    if (history.length === 0 && schemaFeedback) {
      firstParts.push(`Schema feedback: ${JSON.stringify(schemaFeedback)}`);
    }

    const firstMsg =
      firstParts.length > 0
        ? firstParts.join("\n")
        : JSON.stringify(workingMemory, null, 2);

    const messages: Message[] = [{ role: "user", content: firstMsg }];

    for (const turn of history) {
      const action = turn.action ?? {};
      const actionType = action.action_type;
      const turnNum = turn.turn ?? 0;

      if (actionType === "tool_call") {
        const toolUseId = `toolu_history_${turnNum}`;
        // Native Anthropic tool_use block in assistant message
        const assistantContent = [
          { type: "text", text: `[Step ${turnNum}]` },
          {
            type: "tool_use",
            id: toolUseId,
            name: action.tool_name ?? "unknown",
            input: action.arguments ?? {},
          },
        ];
        messages.push({ role: "assistant", content: assistantContent });

        // Native tool_result block in user message
        // Bug C fix: use structured JSON for tool result content
        const toolResultData = turn.tool_result;
        let resultContent: string;
        if (toolResultData && typeof toolResultData === "object") {
          resultContent = JSON.stringify(toolResultData);
        } else {
          resultContent = String(turn.observation ?? "");
        }

        messages.push({
          role: "user",
          content: [
            {
              type: "tool_result",
              tool_use_id: toolUseId,
              content: resultContent,
            },
          ],
        });
      } else if (actionType === "final_response") {
        // Bug A fix: insert bridging user message if last message is also assistant
        if (messages.length > 0 && messages[messages.length - 1].role === "assistant") {
          messages.push({ role: "user", content: "Acknowledged. Continue." });
        }
        messages.push({
          role: "assistant",
          content: `[Step ${turnNum}] ${action.content ?? ""}`,
        });
      }
    }

    // If history was non-empty and last message is assistant, add continuation prompt
    if (history.length > 0 && messages[messages.length - 1].role === "assistant") {
      let continuation = "Continue with the next step.";
      if (schemaFeedback) {
        continuation =
          `Schema feedback: ${JSON.stringify(schemaFeedback)}\n` + "Please try again.";
      }
      messages.push({ role: "user", content: continuation });
    }

    return messages;
  }

  static _parseResponse(response: any): Record<string, unknown> {
    let result: Record<string, unknown> = {};
    let textContent: string | null = null;

    for (const block of response.content) {
      if (block.type === "tool_use") {
        result = {
          type: "tool_call",
          tool_name: block.name,
          arguments: block.input,
        };
        break; // tool_use takes priority
      }
      if (block.type === "text") {
        textContent = block.text;
      }
    }

    if (Object.keys(result).length === 0 && textContent !== null) {
      result = { type: "final_response", content: textContent };
    } else if (Object.keys(result).length === 0) {
      throw new Error("Anthropic response contained no usable content blocks.");
    }

    // Extract usage information
    if (response.usage) {
      result._usage = {
        input_tokens: response.usage.input_tokens,
        output_tokens: response.usage.output_tokens,
        total_tokens: response.usage.input_tokens + response.usage.output_tokens,
      };
    }

    return result;
  }

  private static _resolveApiKey(explicitKey?: string): string {
    const key = explicitKey || (process.env.ANTHROPIC_API_KEY ?? "").trim();
    if (key) return key;

    // Try .env file
    const envPath = path.resolve(__dirname, "../../.env");
    try {
      const envContent = fs.readFileSync(envPath, "utf-8");
      for (const line of envContent.split("\n")) {
        const stripped = line.trim();
        if (!stripped || stripped.startsWith("#") || !stripped.includes("=")) continue;
        const [name, ...rest] = stripped.split("=");
        if (name.trim() === "ANTHROPIC_API_KEY") {
          return rest.join("=").trim().replace(/^["']|["']$/g, "");
        }
      }
    } catch {
      // .env file doesn't exist, continue
    }

    throw new Error(
      "Anthropic API key not found. Set ANTHROPIC_API_KEY environment variable or pass apiKey explicitly.",
    );
  }
}
